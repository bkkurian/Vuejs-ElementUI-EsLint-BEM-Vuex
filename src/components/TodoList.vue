<template>
  <div class="todolist">
   <div  class="el-row todolist__mainContent" > 
      <div class="el-row">
        <div class="el-col el-col-20">
          <el-input placeholder="Create task" name="addTodoItem"  v-model="newTodo" @keyup.enter.native="addTodo()"></el-input> 
        </div>
        <div class="el-col el-col-1">&nbsp;
        </div>
        <div class="el-col el-col-3">
              <el-button type="button" name="AddButton"  size="medium" class="el-button--primary" @click="addTodo()"><span>Add</span></el-button>
        </div>
      </div>
    </div>
    <transition-group>
      <div v-for="(todo) in todosFiltered" :key="todo.id" class="transitiongroup__todoItem">
        <div class="transitiongroup__todoItem__left">
          <el-checkbox  type="checkbox" name="completedCheckbox" v-model="todo.completed" @change="completeTodo(todo)"></el-checkbox>
          <div class="transitiongroup__todoItem__left__label" :class="{ completed : todo.completed }">{{ todo.title }}</div>
        </div>
        <div class="transitiongroup__todoItem__removeItem" @click="removeTodo(todo._id)">
          &times;
        </div>
      </div>
    </transition-group>
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';

export default {
  name: 'todo-list',
  data() {
    return this.$store.state.list;
  },
  computed: mapGetters(['todosFiltered']),
  methods: mapActions(['addTodo', 'removeTodo', 'completeTodo', 'loadTodo']),
  mounted() {
    this.$store.dispatch('loadTodo');
  },
};
</script>

<style lang="scss">
.todolist {
  max-width: 800px;
  margin: 0 auto;
  border: 1px solid lightgray;
  background: white;
  border-radius: 3px;
}
.todolist__mainContent {
  border-bottom: 1px solid lightgray;
  padding: 1em;
}
.transitiongroup__todoItem {
  padding-left: 1em;
  padding-right: 1em;
  border-bottom: 1px solid #eff6fc;
  display: flex;
  align-items: center;
  justify-content: space-between;
  animation-duration: 0.3s;
  font-size: 12px !important;
  &:hover {
    background: #eff6fc;
  }
}


.transitiongroup__todoItem__removeItem {
  cursor: pointer;
  margin-left: 14px;
  &:hover {
    color: black;
  }
}

.transitiongroup__todoItem__left {
  display: flex;
  align-items: center;
}

.transitiongroup__todoItem__left__label {
  padding: 10px;
  margin-left: 12px;
}

.completed {
  text-decoration: bold;
  color: black;
}
</style>
