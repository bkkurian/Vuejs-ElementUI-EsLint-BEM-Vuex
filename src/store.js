import Vue from 'vue';
import Vuex from 'vuex';
import axios from 'axios';

Vue.use(Vuex);
const storeData = {
    newTodo: '',
    idForTodo: 1,
    beforeEditCache: '',
    todos: [],
};
const getters = {
    todosFiltered: (state) => {
        return state.list.todos;
    },
};
const mutations = {
    addTodo(state) {
        const newTodo = {
            id: state.list.idForTodo,
            title: state.list.newTodo,
            completed: false,
        };
        axios.post('http://localhost:8081/todos/', newTodo)
            .then((res) => {
                console.log('Added', res);
                state.list.todos.push(res.data.createdTodo);
                state.list.newTodo = '';
                state.list.idForTodo = state.list.idForTodo + 1;
                return res;
            });
    },
    removeTodo(state, { index }) {
        // state.list.todos.splice(index, 1);
        console.log(index);
        const todosUpdatedList = [];
        axios.delete(`http://localhost:8081/todos/${index}`)
            .then((res) => {
                state.list.todos.forEach((todo) => {
                    if (todo._id !== index) { todosUpdatedList.push(todo); }
                });
                state.list.todos = todosUpdatedList;
                return res;
            });
    },
    completeTodo(state, { todoSelected }) {
        state.list.todos.forEach((todo) => {
            if (todo.id === todoSelected.id) { todo.completed = todoSelected.completed; }
        });
        axios.patch(`http://localhost:8081/todos/${todoSelected._id}`, todoSelected)
            .then((res) => {
                return res;
            });
    },
    loadTodo(state) {
        axios.get('http://localhost:8081/todos/')
            .then((response) => {
                state.list.todos = response.data;
                state.list.newTodo = '';
                state.list.idForTodo = (response.data[response.data.length - 1].id) + 1;
                console.log(state);
            })
            .catch((e) => {
                console.log('Error Occured', e);
            });
    },
};
const actions = {
    addTodo: ({ commit }) => {
        commit('addTodo');
    },
    removeTodo: ({ commit }, index) => {
        commit('removeTodo', { index });
    },
    completeTodo: ({ commit }, todoSelected) => {
        commit('completeTodo', { todoSelected });
    },
    loadTodo: ({ commit }, todoSelected) => {
        commit('loadTodo', { todoSelected });
    },
};

export const store = new Vuex.Store({
    state: {
        list: storeData,
    },
    getters,
    mutations,
    actions,
});
export default store;
