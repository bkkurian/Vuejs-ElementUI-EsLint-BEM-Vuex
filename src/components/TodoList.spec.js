import { mount, shallowMount } from '@vue/test-utils'
import TodoList from './TodoList.vue'
import store from '../store'
import Vue from 'vue'
import ElementUI from 'element-ui';
Vue.config.productionTip = false
Vue.use(ElementUI);
const state = {
  list: {
    newTodo: '',
    idForTodo: 1,
    beforeEditCache: '',
    todos: []
  }
}
const getters = {
  todosFiltered: state => state.list.todos
}
// We can write tests using describe and it
describe('TodoList.vue', () => {
  test('Renders Element UI', () => {
    const wrapper = shallowMount(TodoList, { store })
    expect(wrapper.contains('elinput-stub')).toBe(true)
  })
})
it('TodoList snapshot should match with default snapshot', () => {
  const wrapper = shallowMount(TodoList, { store })
  expect(wrapper.element).toBeDefined();
  expect(wrapper.element).toMatchSnapshot();
});

it('TodoList should render ADD Button ', () => {
  const wrapper = shallowMount(TodoList, { store })
  expect(wrapper.find('span').text()).toBe('Add');
});
