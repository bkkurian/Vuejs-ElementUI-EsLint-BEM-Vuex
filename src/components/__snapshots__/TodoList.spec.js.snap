// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList snapshot should match with default snapshot 1`] = `
<div
  class="todolist"
>
  <div
    class="el-row todolist__mainContent"
  >
    <div
      class="el-row"
    >
      <div
        class="el-col el-col-20"
      >
        <elinput-stub
          autocomplete="off"
          name="addTodoItem"
          placeholder="Create task"
          type="text"
          validateevent="true"
          value=""
        />
      </div>
       
      <div
        class="el-col el-col-1"
      >
        Â 
      
      </div>
       
      <div
        class="el-col el-col-3"
      >
        <elbutton-stub
          class="el-button--primary"
          icon=""
          name="AddButton"
          nativetype="button"
          size="medium"
          type="button"
        >
          <span>
            Add
          </span>
        </elbutton-stub>
      </div>
    </div>
  </div>
   
  <span />
</div>
`;
